Prepared statement parameters in lambda expressions throws exception

See the following example query and error
```
presto> prepare stat from select filter(array[1, 2, 3], x -> x > ?);
PREPARE               

presto> execute stat using 2;

Query 20210426_192410_20062_hw97y failed: line 1:1: Incorrect number of parameters: expected 0 but found 1execute stat using 2
```


This occurs because the DefaultTraversalVisitor (which we use to gather all the parameters) doesn't have an entry for visitLambdaExpression, so we never traverse the body of the lambda expression.

However, if you add an entry for visitLambdaExpression() to DefaultTraversalVisitor, you get the following error later in planning even without the prepared statement:

```
java.lang.AssertionError: Execution of 'actual' query failed: select filter(array[1, 2, 3], x -> x > 2)

	at org.testng.Assert.fail(Assert.java:83)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:151)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:107)
	at com.facebook.presto.tests.AbstractTestQueryFramework.assertQuery(AbstractTestQueryFramework.java:155)
	at com.facebook.presto.tests.AbstractTestQueries.testExecuteUsingLambda(AbstractTestQueries.java:4810)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at org.testng.TestRunner.privateRun(TestRunner.java:756)
	at org.testng.TestRunner.run(TestRunner.java:610)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
	at org.testng.TestNG.runSuites(TestNG.java:1133)
	at org.testng.TestNG.run(TestNG.java:1104)
	at com.intellij.rt.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:66)
	at com.intellij.rt.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:109)
Caused by: java.lang.RuntimeException: Invalid node. Expression dependencies ([expr]) not in source plan output ([])
	at com.facebook.presto.tests.AbstractTestingPrestoClient.execute(AbstractTestingPrestoClient.java:124)
	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:546)
	at com.facebook.presto.tests.DistributedQueryRunner.execute(DistributedQueryRunner.java:514)
	at com.facebook.presto.tests.QueryAssertions.assertQuery(QueryAssertions.java:148)
	... 27 more
Caused by: java.lang.IllegalArgumentException: Invalid node. Expression dependencies ([expr]) not in source plan output ([])
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:163)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.checkDependencies(ValidateDependenciesChecker.java:737)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitProject(ValidateDependenciesChecker.java:287)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitProject(ValidateDependenciesChecker.java:104)
	at com.facebook.presto.spi.plan.ProjectNode.accept(ProjectNode.java:93)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitProject(ValidateDependenciesChecker.java:276)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitProject(ValidateDependenciesChecker.java:104)
	at com.facebook.presto.spi.plan.ProjectNode.accept(ProjectNode.java:93)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitProject(ValidateDependenciesChecker.java:276)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitProject(ValidateDependenciesChecker.java:104)
	at com.facebook.presto.spi.plan.ProjectNode.accept(ProjectNode.java:93)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitProject(ValidateDependenciesChecker.java:276)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitProject(ValidateDependenciesChecker.java:104)
	at com.facebook.presto.spi.plan.ProjectNode.accept(ProjectNode.java:93)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitOutput(ValidateDependenciesChecker.java:332)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker$Visitor.visitOutput(ValidateDependenciesChecker.java:104)
	at com.facebook.presto.sql.planner.plan.OutputNode.accept(OutputNode.java:84)
	at com.facebook.presto.sql.planner.plan.InternalPlanNode.accept(InternalPlanNode.java:33)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker.validate(ValidateDependenciesChecker.java:101)
	at com.facebook.presto.sql.planner.sanity.ValidateDependenciesChecker.validate(ValidateDependenciesChecker.java:96)
	at com.facebook.presto.sql.planner.sanity.PlanChecker.lambda$validateIntermediatePlan$1(PlanChecker.java:88)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:407)
	at com.facebook.presto.sql.planner.sanity.PlanChecker.validateIntermediatePlan(PlanChecker.java:88)
	at com.facebook.presto.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:186)
	at com.facebook.presto.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:179)
	at com.facebook.presto.execution.SqlQueryExecution.doAnalyzeQuery(SqlQueryExecution.java:408)
	at com.facebook.presto.execution.SqlQueryExecution.analyzeQuery(SqlQueryExecution.java:394)
	at com.facebook.presto.execution.SqlQueryExecution.start(SqlQueryExecution.java:344)
	at com.facebook.presto.$gen.Presto_null__testversion____20210426_214137_4.run(Unknown Source)
	at com.facebook.presto.execution.SqlQueryManager.createQuery(SqlQueryManager.java:254)
	at com.facebook.presto.dispatcher.LocalDispatchQuery.lambda$startExecution$5(LocalDispatchQuery.java:114)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

```

--- a/presto-main/src/main/java/com/facebook/presto/execution/ParameterExtractor.java
+++ b/presto-main/src/main/java/com/facebook/presto/execution/ParameterExtractor.java
@@ -14,6 +14,8 @@
 package com.facebook.presto.execution;
 
 import com.facebook.presto.sql.tree.DefaultTraversalVisitor;
+import com.facebook.presto.sql.tree.LambdaArgumentDeclaration;
+import com.facebook.presto.sql.tree.LambdaExpression;
 import com.facebook.presto.sql.tree.Parameter;
 import com.facebook.presto.sql.tree.Statement;
 
@@ -52,6 +54,16 @@ public class ParameterExtractor
         public Void visitParameter(Parameter node, Void context)
         {
             parameters.add(node);
+            return null;
+        }
+
+        @Override
+        protected Void visitLambdaExpression(LambdaExpression node, Void context)
+        {
+            process(node.getBody(), context);
+            for (LambdaArgumentDeclaration argument : node.getArguments()) {
+                process(argument, context);
+            }
             return null;
         }
     }

--- a/presto-spark-base/src/test/java/com/facebook/presto/spark/TestPrestoSparkAbstractTestQueries.java
+++ b/presto-spark-base/src/test/java/com/facebook/presto/spark/TestPrestoSparkAbstractTestQueries.java
@@ -175,6 +175,12 @@ public class TestPrestoSparkAbstractTestQueries
     }
 
     @Override
+    public void testExecuteWithParametersInLambda()
+    {
+        // prepared statement is not supported by Presto on Spark
+    }
+
+    @Override
     public void testExplainDdl()
     {
         // DDL statements are not supported by Presto on Spark

--- a/presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java
+++ b/presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java
@@ -4871,6 +4871,17 @@ public abstract class AbstractTestQueries
     }
 
     @Test
+    public void testExecuteWithParametersInLambda()
+    {
+        String query = "SELECT filter(array[1, 2, 3], x -> x > ?)";
+        Session session = Session.builder(getSession())
+                .addPreparedStatement("my_query", query)
+                .build();
+
+        assertQuery(session, "EXECUTE my_query USING 2", "SELECT array[3]");
+    }
+
+    @Test
     public void testExecuteWithParametersInGroupBy()
     {
         try {
@@ -5725,9 +5736,9 @@ public abstract class AbstractTestQueries
     {
         assertQueryFails(
                 "select y, map_union_sum(x) from (select 1 y, map(array['x', 'z', 'y'], cast(array[null,30,100] as array<tinyint>)) x " +
-                "union all select 1 y, map(array['x', 'y'], cast(array[1,100] as array<tinyint>))x) group by y", ".*Value 200 exceeds MAX_BYTE.*");
+                        "union all select 1 y, map(array['x', 'y'], cast(array[1,100] as array<tinyint>))x) group by y", ".*Value 200 exceeds MAX_BYTE.*");
         assertQueryFails(
                 "select y, map_union_sum(x) from (select 1 y, map(array['x', 'z', 'y'], cast(array[null,30, 32760] as array<smallint>)) x " +
-                "union all select 1 y, map(array['x', 'y'], cast(array[1,100] as array<smallint>))x) group by y", ".*Value 32860 exceeds MAX_SHORT.*");
+                        "union all select 1 y, map(array['x', 'y'], cast(array[1,100] as array<smallint>))x) group by y", ".*Value 32860 exceeds MAX_SHORT.*");
     }
 }

